README.md
# Yearn Web Lib
![](./.github/og.jpeg)

Yearn web Lib is a library of standard components used through Yearn's Projects.    
This library is made for React projects with the idea to be light, efficient and easy to use.  
We are using React + Tailwindcss + ethersjs for the web3 package, and some contexts are available to correctly wrap your app.

Please check @yearn/web-template for documentation and usage.

The stack used for this project is the following:
- 🚀 [Next](https://nextjs.org) — JavaScript library for user interfaces
- ▲ [Vercel](https://vercel.com) — Cloud deployment platform
- 🏎 [Turborepo](https://turborepo.org) — High-performance build system for Monorepos
- 🛠 [Tsup](https://github.com/egoist/tsup) — TypeScript bundler powered by esbuild
- 📄 [TypeScript](https://www.typescriptlang.org/) for static type checking
- 💄 [ESLint](https://eslint.org/) for code linting
- ⚙️ [GitHub Actions](https://github.com/changesets/action) for fully automated package publishing

### Install
```sh
yarn add @yearn-finance/web-lib
```
This repo is mirrored on [NPM](https://www.npmjs.com/package/@yearn-finance/web-lib)

### Useful Commands
- `yarn build` - Build all packages including the Storybook site
- `yarn dev` - Run all packages locally and preview with Storybook
- `yarn lint` - Lint all packages
- `yarn changeset` - Generate a changeset
- `yarn clean` - Clean up all `node_modules` and `dist` folders (runs each package's clean script)


### Apps & Packages
This Turborepo includes the following packages and applications:

- `apps/docs`: Component documentation site, also playground
- `packages/web-lib`: Actual library for Yearn's projects

Each package and app is 100% [TypeScript](https://www.typescriptlang.org/). Yarn Workspaces enables us to "hoist" dependencies that are shared between packages to the root `package.json`. This means smaller `node_modules` folders and a better local dev experience. To install a dependency for the entire monorepo, use the `-W` workspaces flag with `yarn add`.


### Releasing
When you push your code to GitHub, the [GitHub Action](https://github.com/changesets/action) will run the publish script will run. This will:
- Deploy the `app/dosc` app to [Vercel](https://vercel.com)
- Publish the `@yearn/web-lib` package to the Github Registry
- Publish the `@yearn-finance/web-lib` package to [npm](https://www.npmjs.com/)

In order to trigger a new version of the web-lib, the commit message **MUST** start with one of the following:
- `patch:` - This will trigger a new patch version of the web-lib
- `minor:` - This will trigger a new minor version of the web-lib
- `major:` - This will trigger a new major version of the web-lib

We recommand you to use [bump](https://github.com/JS-DevTools/version-bump-prompt) to generate a new version number for each release, with the following commands:
```bash
bump -a -p -c "patch: " #for a patch version, [-a] is for [git commit -a], [-p] is for [git push] and [-c] is for [git commit -m].
bump -a -p -c "minor: " #for a minor version, [-a] is for [git commit -a], [-p] is for [git push] and [-c] is for [git commit -m].
bump -a -p -c "major: " #for a major version, [-a] is for [git commit -a], [-p] is for [git push] and [-c] is for [git commit -m].
```

Release will only occurs on `main` branch, only via an authorized member of Yearn of via the Github Action.

### How to setup

#### Setup tsconfig
With TS config, you should add some paths to be sure to correctly link the web lib: 
```ts
	"baseUrl": ".",
	"paths": {
		"@yearn-finance/web-lib/components": ["./node_modules/@yearn-finance/web-lib/dist/components"],
		"@yearn-finance/web-lib/layouts": ["./node_modules/@yearn-finance/web-lib/dist/layouts"],
		"@yearn-finance/web-lib/contexts": ["./node_modules/@yearn-finance/web-lib/dist/contexts"],
		"@yearn-finance/web-lib/hooks": ["./node_modules/@yearn-finance/web-lib/dist/hooks"],
		"@yearn-finance/web-lib/icons": ["./node_modules/@yearn-finance/web-lib/dist/icons"],
		"@yearn-finance/web-lib/utils": ["./node_modules/@yearn-finance/web-lib/dist/utils"],
	},
```

#### Import the CSS
Create a default `style.css` file in your project root, and add that in it:
```scss
/* This will load Tailwindcss + all the overwrite from Yearn lib */
@import '@yearn-finance/web-lib/dist/style.css';
```

Then, setup your `tailwind.config.js` file to enable detection of your style and prod optimization:
```js
const {join} = require('path');
module.exports = {
	presets: [
		require('@yearn-finance/web-lib/tailwind.plugin')
	],
	content: [
		join(__dirname, 'pages', '**', '*.{js,jsx,ts,tsx}'),
		join(__dirname, 'components', 'icons', '**', '*.{js,jsx,ts,tsx}'),
		join(__dirname, 'components', 'logo', '**', '*.{js,jsx,ts,tsx}'),
		join(__dirname, 'components', 'strategies', '**', '*.{js,jsx,ts,tsx}'),
		join(__dirname, 'components', 'vaults', '**', '*.{js,jsx,ts,tsx}'),
		join(__dirname, 'components', '**', '*.{js,jsx,ts,tsx}'),
		join(__dirname, 'node_modules', '@yearn-finance', 'web-lib', 'dist', 'layouts', '**', '*.js'),
		join(__dirname, 'node_modules', '@yearn-finance', 'web-lib', 'dist', 'components', '**', '*.js'),
		join(__dirname, 'node_modules', '@yearn-finance', 'web-lib', 'dist', 'contexts', '**', '*.js'),
		join(__dirname, 'node_modules', '@yearn-finance', 'web-lib', 'dist', 'icons', '**', '*.js'),
		join(__dirname, 'node_modules', '@yearn-finance', 'web-lib', 'dist', 'utils', '**', '*.js')
	],
	.....
};
```


#### Setup the env
Ensure your env are set. Here is the list of the stuff to set:
```bash
WEBSITE_URI: 'https://my-web3-app.major.farm',
WEBSITE_NAME: 'My awesome yWeb3 app',
WEBSITE_TITLE: 'My awesome yWeb3 app',
WEBSITE_DESCRIPTION: 'Welcome to my awesome Yearn Web3 app. This is a super description that will be used for the SEO stuffs',
PROJECT_GITHUB_URL: 'https://github.com/me/yweb3-awesome',

USE_WALLET: true,
USE_PRICES: false,
USE_PRICE_TRI_CRYPTO: false,
USE_FEEDBACKS: true,

CG_IDS: ['yearn-finance'],
TOKENS: [['0x0bc529c00C6401aEF6D220BE8C6Ea1667F6Ad93e', 18, 1]],

WEB_SOCKET_URL: {
	1: process.env.WS_URL_MAINNET,
	250: process.env.WS_URL_FANTOM,
	42161: process.env.WS_URL_ARBITRUM
},
JSON_RPC_URL: {
	1: process.env.RPC_URL_MAINNET,
	250: process.env.RPC_URL_FANTOM,
	42161: process.env.RPC_URL_ARBITRUM
},
ALCHEMY_KEY: process.env.ALCHEMY_KEY,
INFURA_KEY: process.env.INFURA_KEY
```


## How to use
Usage is way simpler. You first need to wrap you app with the WithYearn context, and then you can use the components from the library.
```tsx
import	{WithYearn}		from	'@yearn-finance/web-lib/contexts';

function	MyApp(props: AppProps): ReactElement {
	const	{Component, pageProps} = props;
	
	return (
		<WithYearn>
			<AppWrapper
				Component={Component}
				pageProps={pageProps} />
		</WithYearn>
	);
}
```


CHANGELOG.md
# Releases

## @yearn-finance/web-lib@0.8.9 to @yearn-finance/web-lib@0.8.17
- [9df6ee9f191905ca60ca127da83af9fbd4b2d021](https://github.com/yearn/web-lib/commit/9df6ee9f191905ca60ca127da83af9fbd4b2d021): Feat: Deprecate `@yearn/web-lib` in favor of `@yearn-finance/web-lib` by disabling Github Actions.
- [9f9ac258c3b4af275330e7053089750ff33032ca](https://github.com/yearn/web-lib/commit/9f9ac258c3b4af275330e7053089750ff33032ca): Fix: Fix `TxHashWithActions` to use the correct URI
- [ab5b62db2691650fcd0d5627b0b4e9921cc09500](https://github.com/yearn/web-lib/commit/ab5b62db2691650fcd0d5627b0b4e9921cc09500): Fix: Removing auto-changeset actions.
- [3d4a83cfc7f5cad53e676fb193ff3227cfb3a2bf](https://github.com/yearn/web-lib/commit/3d4a83cfc7f5cad53e676fb193ff3227cfb3a2bf): Fix: Export the `style.css` file correctly (enable `postcss`)

## @yearn-finance/web-lib@0.8.8
- [eb0fe2f16e0fce4b79798c72b315ef764b446754](https://github.com/yearn/web-lib/commit/eb0fe2f16e0fce4b79798c72b315ef764b446754): Feat: Add the `partnerContractAddress` to the default Networks settings.
- [eb0fe2f16e0fce4b79798c72b315ef764b446754](https://github.com/yearn/web-lib/commit/eb0fe2f16e0fce4b79798c72b315ef764b446754): Fix: Set the default ChainID to the one sent as `options.defaultChainID` (default: 1)

## @yearn-finance/web-lib@0.8.7
- [eb0fe2f16e0fce4b79798c72b315ef764b446754](https://github.com/yearn/web-lib/commit/eb0fe2f16e0fce4b79798c72b315ef764b446754): Fix: Set the ENS listener to mainnet only instead of default network.

## @yearn-finance/web-lib@0.8.4 to @yearn-finance/web-lib@0.8.6
- [e09b40bdd2b99aacf1589e3c428076ab09961981](https://github.com/yearn/web-lib/commit/e09b40bdd2b99aacf1589e3c428076ab09961981): Feat: Add a new CSS variable `--default-rounded` to define the default border radius, available as a class via `rounded-default`
- [e09b40bdd2b99aacf1589e3c428076ab09961981](https://github.com/yearn/web-lib/commit/e09b40bdd2b99aacf1589e3c428076ab09961981): Fix: Remove extra font import and some color failure
- [e09b40bdd2b99aacf1589e3c428076ab09961981](https://github.com/yearn/web-lib/commit/e09b40bdd2b99aacf1589e3c428076ab09961981): Feat: Split the theming system in multiple CSS files to be able to only import the required ones:
	- `style.css` will import `style.light.css` and `style.dark.css`
	- `style.light.css` define the colors and specific styles for the light theme
	- `style.dark.css` define the colors and specific styles for the dark theme
	- `style.macarena.css` define the colors and specific styles for the macarena theme


## @yearn-finance/web-lib@0.8.0 to @yearn-finance/web-lib@0.8.3
- [15cc1e9901fb706d8e5a34733080c61352b4a32f](https://github.com/yearn/web-lib/commit/15cc1e9901fb706d8e5a34733080c61352b4a32f): Doc: Add an OG image to the Readme
- [fdb6cbd0d37df37abc3430b82dcaba3c2ee16e75](https://github.com/yearn/web-lib/commit/fdb6cbd0d37df37abc3430b82dcaba3c2ee16e75): Fix: Remove extra precision for nested CSS class to avoid conflicts.
- [75ce536e367e45f4c25008237e2b3990fb783865](https://github.com/yearn/web-lib/commit/75ce536e367e45f4c25008237e2b3990fb783865): Fix: Add more precision over the CSS classes to avoid conflicts.
- [e5c19f9c9a7185ce57f35f9508f0ccb65a639986](https://github.com/yearn/web-lib/commit/e5c19f9c9a7185ce57f35f9508f0ccb65a639986): Fix: Replace deprecated color for the modals from (`bg-dark` to `bg-black`).
- [0d4a52d257d17cb33f62cf47609ec419d1e62af7](https://github.com/yearn/web-lib/commit/0d4a52d257d17cb33f62cf47609ec419d1e62af7): Fix: Linting the code.

## @yearn-finance/web-lib@0.8.0
- [335891d732c35b04b516d4eb2762dc8eed3e4919](https://github.com/yearn/web-lib/commit/335891d732c35b04b516d4eb2762dc8eed3e4919): Update the color system  
> **Warning**  
> This bump is a breaking change about the color names.  
> Please, update at least the following declaration `color-(bg|text|border)-(all-previous-names)` to the current new HSL color names.  
> See bellow for mapping.  

Color system update, updating color system from RGB to HSL.  
HEX and RGB are the most common formats for representing color on the web, but they’re not the most useful.  
Using HEX or RGB, colors that have a lot in common visually look nothing alike in code.  
HSL fixes this by representing colors using attributes the human-eye intuitively perceives: hue, saturation, and lightness.  
Hue is a color’s position on the color wheel — it’s the attribute of a color that lets us identify two colors as “blue” even if they aren’t identical.  

<details>
  <summary>HSL vs RGB</summary>
  
    :root {
    	--color-background: rgb(244 247 251),
    	--color-background-variant: rgb(224 234 255),
    	--color-surface: rgb(255 255 255),
    	--color-surface-variant: rgb(249 251 253),
    	--color-primary: rgb(6 87 249),
    	--color-primary-variant: rgb(0 74 223),
    	--color-secondary: rgb(224 234 255),
    	--color-secondary-variant: rgb(198 215 249),
    	--color-disabled: rgb(206 213 227),
    	--color-dark: rgb(20 20 20)
    }

    :root {
    	--color-neutral-0: hsl(0 0% 100%),
    	--color-neutral-100: hsl(220 50% 98%),
    	--color-neutral-200: hsl(220 47% 97%),
    	--color-neutral-300: hsl(220 100% 94%),
    	--color-neutral-400: hsl(220 27% 85%),
    	--color-neutral-500: hsl(220 22% 36%),
    	--color-neutral-600: hsl(220 22% 25%),
    	--color-neutral-700: hsl(220 100% 14%),
    	--color-neutral-800: hsl(0 0% 8%),
    	--color-neutral-900: hsl(0 0% 0%),
    }
</details>

<details>
  <summary>New color names</summary>

    '--color-neutral-0'    =>  '--color-surface' | '--color-button-filled-text' | '--color-button-outlined-primary'
    '--color-neutral-100'  =>  '--color-surface-variant'
    '--color-neutral-200'  =>  '--color-background' | '--color-button-disabled-primary'
    '--color-neutral-300'  =>  '--color-background-variant'
    '--color-neutral-400'  =>  '--color-disabled' | '--color-icons-primary' | '--color-typo-off' | '--color-disabled-text'

next.config.js
const Dotenv = require('dotenv-webpack');

module.exports = ({
	experimental: {
		concurrentFeatures: true
	},
	plugins: [new Dotenv()],
	images: {
		domains: [
			'rawcdn.githack.com'
		]
	},
	env: {
		/* 🔵 - Yearn Finance **************************************************
		** Stuff used for the SEO or some related elements, like the title, the
		** github url etc.
		**********************************************************************/
		WEBSITE_URI: 'https://yearn.systems/',
		WEBSITE_NAME: 'yWeb',
		WEBSITE_TITLE: 'yWeb',
		WEBSITE_DESCRIPTION: 'Template used for Yearn\'s projects',
		PROJECT_GITHUB_URL: 'https://github.com/yearn/web-lib',

		/* 🔵 - Yearn Finance **************************************************
		** Some config used to control the behaviour of the web library. By
		** default, all of theses are set to false.
		** USE_WALLET: should we allow the user to connect a wallet via
		**             metamask or wallet connect?
		** USE_PRICES: should we fetch the prices for a list of tokens? If true
		**             the CG_IDS array should be populated with the tokens
		**             to fetch.
		** USE_PRICE_TRI_CRYPTO: should we fetch the special Tri Crypto token
		** 			   price? (require blockchain call)
		** USE_FEEDBACKS: should we enable the feedback button?
		**********************************************************************/
		USE_WALLET: true,
		USE_PRICES: true,
		USE_PRICE_TRI_CRYPTO: false,
		USE_FEEDBACKS: true,
		USE_NETWORKS: true,
		CG_IDS: ['yearn-finance'],
		TOKENS: [
			['0x0bc529c00C6401aEF6D220BE8C6Ea1667F6Ad93e', 18, 1]
		],
		RPC_URL: {
			1: process.env.RPC_URL_MAINNET,
			250: process.env.RPC_URL_FANTOM || 'https://rpc.ftm.tools',
			42161: process.env.RPC_URL_ARBITRUM || 'https://arbitrum.public-rpc.com'
		},
		ALCHEMY_KEY: process.env.ALCHEMY_KEY,

		FEEBACKS_TYPE: 'github',
		LINEAR_OAUTH_TOKEN: process.env.LINEAR_OAUTH_TOKEN,
		LINEAR_TEAM_ID: process.env.LINEAR_TEAM_ID,
		LINEAR_PROJECT_NAME: process.env.LINEAR_PROJECT_NAME,

		GITHUB_AUTH_TOKEN: process.env.GITHUB_AUTH_TOKEN,
		GITHUB_PROJECT_OWNER: process.env.GITHUB_PROJECT_OWNER,
		GITHUB_PROJECT_REPO: process.env.GITHUB_PROJECT_REPO

	}
});
